lib/vfs/src/host_fs.rs

```diff
@@ -0,0 +1,648 @@
+    fn seek(&mut self, pos: io::SeekFrom) -> io::Result<u64> {
+        self.inner.seek(pos)
+    }
+}
+impl Write for HostFile {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.inner.write(buf)
+    }
+    fn flush(&mut self) -> io::Result<()> {
+        self.inner.flush()
+    }
+    fn write_all(&mut self, buf: &[u8]) -> io::Result<()> {
+        self.inner.write_all(buf)
+    }
+    fn write_fmt(&mut self, fmt: ::std::fmt::Arguments) -> io::Result<()> {
+        self.inner.write_fmt(fmt)
+    }
+}
+
+#[cfg_attr(feature = "enable-serde", typetag::serde)]
+impl VirtualFile for HostFile {
+    fn last_accessed(&self) -> u64 {
+        self.metadata()
+            .accessed()
+            .ok()
+            .and_then(|ct| ct.duration_since(SystemTime::UNIX_EPOCH).ok())
+            .map(|ct| ct.as_nanos() as u64)
+            .unwrap_or(0)
+    }
+
+    fn last_modified(&self) -> u64 {
+        self.metadata()
+            .modified()
+            .ok()
+            .and_then(|ct| ct.duration_since(SystemTime::UNIX_EPOCH).ok())
+            .map(|ct| ct.as_nanos() as u64)
+            .unwrap_or(0)
+    }
+
+    fn created_time(&self) -> u64 {
+        self.metadata()
+            .created()
+            .ok()
+            .and_then(|ct| ct.duration_since(SystemTime::UNIX_EPOCH).ok())
+            .map(|ct| ct.as_nanos() as u64)
+            .unwrap_or(0)
+    }
+
+    fn size(&self) -> u64 {
+        self.metadata().len()
+    }
+
+    fn set_len(&mut self, new_size: u64) -> Result<(), FsError> {
+        fs::File::set_len(&self.inner, new_size).map_err(Into::into)
+    }
+
+    fn unlink(&mut self) -> Result<(), FsError> {
+        std::fs::remove_file(&self.host_path).map_err(Into::into)
+    }
+    fn sync_to_disk(&self) -> Result<(), FsError> {
+        self.inner.sync_all().map_err(Into::into)
+    }
+
+    fn bytes_available(&self) -> Result<usize, FsError> {
+        // unwrap is safe because of get_raw_fd implementation
+        let host_fd = self.get_raw_fd().unwrap();
+
+        host_file_bytes_available(host_fd)
+    }
+
+    #[cfg(unix)]
+    fn get_raw_fd(&self) -> Option<i32> {
+        use std::os::unix::io::AsRawFd;
+        Some(self.inner.as_raw_fd())
+    }
+    #[cfg(not(unix))]
+    fn get_raw_fd(&self) -> Option<i32> {
+        unimplemented!(
+            "HostFile::get_raw_fd in VirtualFile is not implemented for non-Unix-like targets yet"
+        );
+    }
+}
+
+#[cfg(unix)]
+fn host_file_bytes_available(host_fd: i32) -> Result<usize, FsError> {
+    let mut bytes_found = 0 as libc::c_int;
+    let result = unsafe { libc::ioctl(host_fd, libc::FIONREAD, &mut bytes_found) };
+
+    match result {
+        // success
+        0 => Ok(bytes_found.try_into().unwrap_or(0)),
+        libc::EBADF => Err(FsError::InvalidFd),
+        libc::EFAULT => Err(FsError::InvalidData),
+        libc::EINVAL => Err(FsError::InvalidInput),
+        _ => Err(FsError::IOError),
+    }
+}
+
+#[cfg(not(unix))]
```

Comment:
```suggestion
#[cfg(windows)]
```

Reply:
See https://github.com/wasmerio/wasmer/pull/2491#discussion_r689570334. tl;dr it's fixed by https://github.com/wasmerio/wasmer/pull/2524.